### ISSUE TITLE
Bloc listeners don't unsubscribe from bloc stream when widget is disposed

### ISSUE THREAD

**Issue by ruslic19** [Iteo/hooked_bloc#70]
Hi, there is a bug, which prevents hooked bloc listeners from unsubscribing from bloc stream. The reason of this is that inside of bloc listener hooks, `useMemoized` is used, while it looks like `useEffect` was intented to be used there and because of this mistake, bloc stream subscription is not cancelled when widget is disposed.

So, it means that when we attach listener multiple times to the same long living bloc, listener will be executed multiple times, even though we already disposed widgets which attached those listeners. For example, when listener would be attached inside of dialog which we would open/close multiple times, we would receive this issue.

PR with fix: #69

**Comment by ruslic19** [comment]
Simple app which has this issue reproduced, listener continues to be called even after dialog is closed and then when you reopen dialog, one more listener is attached, so you have 2 calls every second
```
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:flutter_hooks/flutter_hooks.dart';
import 'package:hooked_bloc/hooked_bloc.dart';

void main() {
  runApp(const MyApp());
}

class MyCubit extends Cubit {
  MyCubit() : super(0);

  start() {
    Timer.periodic(const Duration(seconds: 1), (timer) {
      emit(state + 1);
    });
  }
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: BlocProvider(
        create: (_) => MyCubit()..start(),
        child: Builder(builder: (context) {
          return ElevatedButton(
            onPressed: () => showDialog(
              context: context,
              builder: (_) => BlocProvider.value(
                value: BlocProvider.of<MyCubit>(context),
                child: const Dialog(),
              ),
            ),
            child: const Text('Dialog'),
          );
        }),
      ),
    );
  }
}

class Dialog extends HookWidget {
  const Dialog({super.key});

  @override
  Widget build(BuildContext context) {
    final myCubit = useBloc<MyCubit>(closeOnDispose: false);

    useBlocComparativeListener(
      myCubit,
      (bloc, current, context) {
        print('LISTENER CALLED');
      },
      listenWhen: (p, c) => true,
    );

    return ElevatedButton(
      onPressed: () => Navigator.pop(context),
      child: const Text('Close Dialog'),
    );
  }
}
```

**Comment by RSC-88** [comment]
@ruslic19 thx for PR and support. Version with your fixes: 1.4.2
https://pub.dev/packages/hooked_bloc/versions/1.4.2
